### DISCUSSION TITLE
Cron jobs

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cloud#20] [Feature Requests & Enhancements]

**Comment by matthewstick** [comment]
Thank you - please add ◡̈

**Comment by michtio** [comment]
@timkelty any ETA on this, as this is a major blocker in a lot of our projects 😄

**Comment by jamesmacwhite** [comment]
Would love cronjobs as well, however we've been pointed as GitHub actions being an alternative solution for now, until a native feature exists.

https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#schedule

Is there a sample GitHub action that calls the php craft console runner to execute this?

**Reply by timkelty** [reply]
> Is there a sample GitHub action that calls the php craft console runner to execute this?

We can't invoke PHP directly from Github actions, so we need to do it via a web request.

So, you need to create a web controller (protected by an API secret of some sort), and hit that in your Github action via something like:

```
curl https://your-site.com/actions/your-controller \
--header "Authorization: Bearer your-api-key"
```

Then your controller would check the API key, and push a job to the queue to do what you wanted. You could swap the `Authorization` header for whatever form of auth you're comfortable with here (basic auth, etc) – but you presumably don't want the endpoint unprotected.

If it is helpful, I can try and mock up an example – that said, adding proper cronjob support is our top priority feature right now, so shouldn't be far out: https://craftcms.com/roadmap#cloud

**Reply by jamesmacwhite** [reply]
@timkelty Thank you for confirming. I see, I guess with Craft Cloud being serverless that you can't execute PHP in the traditional way. Web request controller would do the job, hope native cronjobs aren't too far way though, because of serverless, the lack of being able to configure an external CI to run a PHP command is a welcome feature.

**Comment by jasonmccallister** [comment]
This feature was released today! 

![CleanShot 2024-12-02 at 15 18 42@2x](https://github.com/user-attachments/assets/b141215a-5c8c-4a61-9c7b-fd15fb3a780a)

**Reply by jamesmacwhite** [reply]
Thanks! I'll look to try it out.

Is the 5 commands limit something that has been set as an initial limit with potential to increase? We have more than 5 cron jobs currently configured in GitHub actions.

Just making sure I'm reading the info correctly, cron jobs through Craft Console are limited to being run on a frequency of 1 hour as a minimum?

**Reply by jasonmccallister** [reply]
> Is the 5 commands limit something that has been set as an initial limit with potential to increase?

Yes, that is our initial limit and we are not sure if/when we will raise it. How many CRON jobs are you currently running for a single environment in GitHub Actions?

> Just making sure I'm reading the info correctly, cron jobs through Craft Console are limited to being run on a frequency of 1 hour as a minimum?

Yes, right now they are limited to a minimum of hourly. We went back and forth on this internally. Do you mind sharing the types of commands and frequency you are running them so we can get better insight to your use case?

**Reply by jamesmacwhite** [reply]
@jasonmccallister Currently, 9 GitHub actions that HTTP triggers to controller. Most are Feed Me feeds, which can be triggered via HTTP no problem as we are just using it as a signal to run a queue job, so I wouldn't necessarily need to make them cronjobs. We have actually just moved away from GitHub actions. There a few issues with using them to send HTTP requests. You have to spin up a container each time, which takes several seconds and under a private repository the usage is not unlimited. I ended up moving to Buddy CI, because we have it for other projects, which is much faster.

The command we want to run polls a server to fetch the latest image and then saved this as an asset for displaying it updates roughly every 10-15 minutes. We have this currently through a queue job, but it is on the line of being suited for a queue job due to the queue being quite busy.

**Comment by matthewstick** [comment]
hi @jasonmccallister 

Awesome.  So the cron jobs we are looking to run are modules and actions.  We have data imports, synchs, etc.  We don't have a need to run craft functions -- like update search index -- on a cron.

So how can I ping these as cron jobs in Craft Cloud:
/actions/dbSynch
/actions/importShowtimes
/api/runBatch

etc.

Thanks!

**Reply by jamesmacwhite** [reply]
@matthewstick Your module can register a console command which is able to be run by Craft CLI. A controller action could also be triggered via a console command too.

**Reply by angrybrad** [reply]
@matthewstick what @jamesmacwhite said. Currently, since cron jobs in Cloud are essentially "scheduled craft commands", anything outside of that would need to be wrapped up and executed via a custom controller action.

**Reply by jamesmacwhite** [reply]
For controllers, the alternative is sending HTTP requests to controller actions to act as triggers, although I'd recommend ensuring that a web controller request doesn't run for too long if you do that, if it's something heavy, I'd recommend putting the main logic into a console command or queue job and just let the HTTP request be the trigger and return 200 OK ASAP to know the trigger was successful.

It's great that native cronjobs are available now though. Not everything is potentially suited to a queue job, hope the 5 jobs/1 hour limit is something that can be changed in time.

**Reply by angrybrad** [reply]
@jamesmacwhite yup - already have plans on how to solve "longer running commands". And we're always open to feedback for things (like 5 job limits)

**Reply by jamesmacwhite** [reply]
@angrybrad The 5 jobs limit, we would currently exceed due to having more than 5 Feed Me feeds. This isn't so much of a problem, by using HTTP requests which run queue jobs via controller. The minimum 1 hour limit of cron jobs, that would be something I'd like to see lowered, perhaps 15 minutes if the backend can deal with it.

I have a use case currently of wanting to trigger a console command every 15 minutes. Currently it is being triggered via HTTP and queue job, but ideally it could do with being executed via console runner, given the queue can be busy and delays the running of it.
